### PR TITLE
Git application

### LINKED ISSUE #126: App to submit PR with annotations into a git repo
It is a very good practice to submit all annotations into a git repository. Now a user should download annotations and after that upload them into a git repository manually. It is necessary to implement a django application which will help with the task.

- [ ] Create new task dialog should have an additional optional field. If a git repository isn't specified it is OK. In this case all works as previously.
- [ ] Support GitLab and GitHub out of the box (use corresponding API to submit PR). Thus need to have a configuration dialog with GitLab/GitHub token, etc... Also GitLab/GitHub access can be setup using the app settings (can we configure app settings using web interface?)
- [ ] Need to have statuses: “all annotations were committed to the git repo” or “there are changes which are not committed yet”
- [ ] Probably it is a good time to implement a status for whole task (completed, in progress, open, etc...). Please review possible django/javascript apps which implement "issue tracker" functionality. Probably there is an open source app which can be integrated into CVAT as is and improve our dashboard.

### PR DESCRIPTION
**Pull request by bsekachev**
Resolved #126

### CODE REVIEW

File: cvat/apps/git/git.py [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+from cvat.apps.engine.plugins import add_plugin
+
+from cvat.apps.git.models import GitData
+from cvat.apps.git.settings import *
+
+import subprocess
+import datetime
+import shutil
+import json
+import git
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat/apps/git/git.py]: @azhavoro 
GitPython only.
It's installed from ```components/git/install.sh```

File: cvat/apps/git/git.py [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+        self.__branch_name = '{}_tid_{}'.format(CVAT_BRANCH_PREFIX, self.__tid)
+        self.__user = user.username
+        self.__cwd = os.path.join(os.getcwd(), "data", str(tid), "repos")
+        self.__diffs_dir = os.path.join(os.getcwd(), "data", str(tid), "repos_diffs")
+
+
+    # Method parses an got URL.
+    # SSH: git@github.com/proj/repos[.git]
+    # HTTP/HTTPS: [http://]github.com/proj/repos[.git]
+    def _parse_url(self):
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat/apps/git/git.py]: @azhavoro 
Done.

**Comment by nmanovic** [comment]
- User for commit should be changed (now it is ExampleOrgz)
- Sync status should be more visible in dashboard. I propose colored background for task item (red - unsync, yellow - sync but not merged, green - merged)

**Comment by nmanovic** [comment]
- diff file was committed into my repository. Is it expected?

**Comment by nmanovic** [comment]
- git url and path isn't dumped into the xml file itself.

**Comment by nmanovic** [comment]
- Need to provide good default value for git path (e.g. `<dump file name>.xml`)

**Comment by bsekachev** [comment]
@nmanovic 

"Error handling if ssh key isn't added should be better. It will be a usual mistake. Need to ask the user to add ssh key into git server explicitly"
Interaction with an user during building is actually good idea?

**Comment by bsekachev** [comment]
@nmanovic 
"diff file was committed into my repository. Is it expected?"
changelog_*.diff commits to repos. It is expected. Other diffs files stores outside repository directory.

**Comment by bsekachev** [comment]
@nmanovic 
"Sync status should be more visible in dashboard. I propose colored background for task item (red - unsync, yellow - sync but not merged, green - merged)"
It is very long for loading because each "get_remote_status" request does request to git server (1-3 sec). I have no idea how implement it effectively yet.

**Comment by bsekachev** [comment]
@nmanovic 
Other comments have been implemented

**Comment by nmanovic** [comment]
> Interaction with an user during building is actually good idea?

It isn't during building. It is when a user try to push something into a repository for which the user doesn't add generated ssh key. It will get some error that the repository cannot be cloned with a lot of extra stuff. I remember that it isn't so easy to understand why a repository cannot be cloned. Let's say that "Probably SSH key was not added to your profile for the cloned repository. Please add the key below and try again".

**Comment by nmanovic** [comment]
> @nmanovic
> "diff file was committed into my repository. Is it expected?"
> changelog_*.diff is commited to repos. It is expected. Other diffs files stores outside repository directory.

Why do we need it?

**Comment by nmanovic** [comment]
> It is very long for loading because each "get_remote_status" request does request to git server (1-3 sec). I have no idea how implement it effectively yet.

Indeed the task is very important. It is really good feature. I need it to be sure that latest tasks are synchronized and it will be inconvenient to click each time on a task to understand its status.

There are several ways to implement that:
- Just cache the latest status always and return it together with task info for dashboard. But it will necessary to click each time the task push as well. But one clicked it will be "save" state.
- Ask users to add a webhook. I don't like the approach but probably it will work best with the approach above.
- Implement a background task which sync the status time to time for non-merged repos and cache it. It will work best from my point of view. Also when a user press "push" button the status is cached as well.

**Comment by bsekachev** [comment]
@nmanovic 
"Why do we need it inside repository? Is it a temporary file for internal purpose?"
No, this contains information about changed objects and dates.

**Comment by nmanovic** [comment]
@bsekachev ,

Is it possible to incorporate the information into git commit message? Why do we need a separate file for that?

**Comment by nmanovic** [comment]
@bsekachev ,

The main problem here that *our users don't need the file*.  You can keep it internally. All relevant information should be added into git log message. Thus it is easy to see history of the repository using `git log` command.

**Comment by nmanovic** [comment]
@bsekachev,

To be clear I don't propose to delete the file. I propose don't commit it and add its content into commit message.

**Comment by bsekachev** [comment]
@nmanovic 
Are that all requirements?

**Comment by nmanovic** [comment]
@bsekachev , Please notify me when it is ready. I have the following error now.

image

**Comment by bsekachev** [comment]
@nmanovic 
You can review it.

**Comment by nmanovic** [comment]
Just one more comment. It seems that after some time it became "red". But after "save" method with non-zero number of bounding boxes it should be "red" immediately.

**Comment by bsekachev** [comment]
@nmanovic 
I have implemented approach with regular status updating how it specified above.
If you check a status manually after changing or merge an annotation, it will be updated. 
Otherwise it will updated after 10 minutes.

Also we have an ability to update status to "!sync" on save event, but we can't do it after merge on a remote repository, because we aren't notified about it.

I'll fix issue with an extension.

**Comment by bsekachev** [comment]
@nmanovic 
Please, review it again.

File: Dockerfile [hunk 3]
```diff
@@ -113,4 +123,4 @@ RUN mkdir data share media keys logs /tmp/supervisord
 RUN python3 manage.py collectstatic
 
 EXPOSE 8080 8443
-ENTRYPOINT ["/usr/bin/supervisord"]
+ENTRYPOINT ["bash", "-c", "eval `/usr/bin/ssh-agent -s` && /usr/bin/supervisord"]
```
[the remark this answers is not in the document]
Comment (nmanovic) [reply Dockerfile]: Look at the example: https://github.com/hauptmedia/docker-ssh-agent/blob/master/supervisord.conf

[the remark this answers is not in the document]
Comment (nmanovic) [reply Dockerfile]: It is possible to specify exact file which will be used for communication using `-a` option. Thus let's specify one more program inside supervisor config and add one explicit variables to all other programs inside the config.

File: ssh/README.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+Put your SSH keys and SSH config here and they will be installed to the CVAT container.
+
+If you have any problems with a git repository cloning inside the CVAT:
+  * Make sure that SSH keys have been added to the CVAT container: echo ```docker exec -it cvat bash -ic 'ls .ssh'```
+  * If you need a proxy for connect to the Internet, specify the socks_proxy variable before build the container. For example:
```
Comment (nmanovic) [inline ssh/README.md]: for connecting?

File: ssh/README.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+```bash
+socks_proxy=proxy-example.com:1080 docker-compose build
+```
+  * Try to clone a repository via SSH direct in the container:
```
Comment (nmanovic) [inline ssh/README.md]: directly?

File: cvat/apps/engine/views.py [hunk 6]
```diff
@@ -104,8 +104,6 @@ def create_task(request):
     return JsonResponse({'tid': db_task.id})
 
 @login_required
-@permission_required(perm=['engine.task.access'],
```
Comment (nmanovic) [inline cvat/apps/engine/views.py]: Need to have extra comment here.

**Pull request merged**